File: lineagelens-backend/app/schemas/agent_trace.py
"""Agent Trace interchange format — cursor/agent-trace spec v0.1.0.

Spec: https://github.com/cursor/agent-trace

Each exported LineageLens record maps to one AgentTraceDocument covering a
single file insertion event.  LineageLens-specific fields (confidence,
workspace_id, evidence, etc.) that have no equivalent in the vendor-neutral
spec are carried in the `metadata` dict under "lineagelens.*" keys, which is
the spec's intended extension mechanism for vendor-specific data.

Field-by-field changes from our pre-spec internal format
("lineagelens-agent-trace/1" → cursor/agent-trace 0.1.0):

  schemaVersion: "lineagelens-agent-trace/1"  → version: "0.1.0" (semver)
  uuid                                         → id
  filePath (flat)                              → files[].path (nested)
  contributorType "AI" (uppercase)             → files[].conversations[].contributor.type "ai" (lowercase)
  model.name                                   → files[].conversations[].contributor.model_id
  lineStart / lineEnd (camelCase, 0-based)     → start_line / end_line (snake_case, 1-indexed, min 1)
  workspaceId                                  → metadata["lineagelens.workspaceId"]
  confidence.score / .level                   → metadata["lineagelens.confidence"]
  evidence                                     → metadata["lineagelens.evidence"]
  insertedCodePreview                          → metadata["lineagelens.insertedCodePreview"]
  netAddedLines                                → metadata["lineagelens.netAddedLines"]
  exportedAt                                   → metadata["lineagelens.exportedAt"]
  tool.adapter/provider/sessionId/...         → metadata["lineagelens.tool"] (slim spec tool has name+version only)

Fields already present that matched (no change needed):
  timestamp (RFC 3339) — same
  tool.name — same location, same semantics
  files[].path — new name for filePath, nested

New spec fields added (populated where possible):
  vcs          — TODO:SPEC: not captured at insertion time; mapped to null
  tool.version — TODO:SPEC: not available in capture data; mapped to null
  content_hash — TODO:SPEC: LineageLens does not compute range hashes; mapped to null

TODO:SPEC The cursor/agent-trace spec is at RFC status (v0.1.0). When it
reaches stable, review whether any new top-level fields absorb our metadata keys.
"""

from __future__ import annotations

from pydantic import BaseModel

SPEC_VERSION = "0.1.0"
# Kept for metadata tagging and backward-compat import detection only.
SCHEMA_VERSION = "lineagelens-agent-trace/1"


class AgentTraceVcs(BaseModel):
    # TODO:SPEC revision is the commit SHA (git), change ID (jj), etc.
    # LineageLens does not capture this at insertion time; mapped to null.
    type: str   # "git" | "jj" | "hg" | "svn"
    revision: str


class AgentTraceTool(BaseModel):
    name: str | None = None
    # TODO:SPEC populate version from the capturing tool's version metadata
    # when available (e.g. Cursor 2.4.0, GitHub Copilot 1.x).
    version: str | None = None


class AgentTraceContributor(BaseModel):
    # Spec enum values are ALL LOWERCASE: "human" | "ai" | "mixed" | "unknown"
    # (Previous internal format used uppercase "AI" — now corrected.)
    type: str   # "human" | "ai" | "mixed" | "unknown"
    # TODO:SPEC model_id should follow models.dev convention, e.g.
    # "anthropic/claude-opus-4-5-20251101". LineageLens stores the raw model
    # name from the proxy/extension which may not be in fully qualified form.
    model_id: str | None = None   # max 250 chars per spec


class AgentTraceRelatedResource(BaseModel):
    type: str
    url: str


class AgentTraceRange(BaseModel):
    start_line: int   # 1-indexed, minimum 1
    end_line: int     # 1-indexed, minimum 1
    # TODO:SPEC content_hash format is "algorithm:hash" (e.g. "murmur3:9f2e8a1b").
    # LineageLens does not currently compute per-range content hashes; null.
    content_hash: str | None = None
    contributor: AgentTraceContributor | None = None


class AgentTraceConversation(BaseModel):
    # TODO:SPEC url should be the conversation lookup URL. LineageLens stores
    # a conversationId string, not a URL; mapped to null.
    url: str | None = None
    contributor: AgentTraceContributor | None = None
    ranges: list[AgentTraceRange]
    related: list[AgentTraceRelatedResource] | None = None


class AgentTraceFile(BaseModel):
    path: str
    conversations: list[AgentTraceConversation]


class AgentTraceDocument(BaseModel):
    """One Agent Trace document per the cursor/agent-trace spec v0.1.0."""

    version: str = SPEC_VERSION   # semver, currently "0.1.0"
    id: str                        # UUID
    timestamp: str                 # RFC 3339
    # TODO:SPEC vcs (commit SHA) is not captured at insertion time; null.
    vcs: AgentTraceVcs | None = None
    tool: AgentTraceTool | None = None
    files: list[AgentTraceFile]
    # Vendor-specific extension bag. LineageLens uses "lineagelens.*" keys.
    metadata: dict | None = None


# Backward-compat alias so import sites that reference AgentTraceRecord
# continue to work without changes.
AgentTraceRecord = AgentTraceDocument

File: lineagelens-base-extension/src/agentTrace.ts
/**
 * Local Agent Trace export for Easy Mode (no backend required).
 *
 * Converts CaptureRecord[] → cursor/agent-trace 0.1.0 JSONL.
 * The field mapping mirrors the Python backend (record_to_agent_trace in
 * lineagelens-backend/app/services/agent_trace_service.py) so that
 * backend-export and extension-export produce identical-shape records.
 *
 * Spec: https://github.com/cursor/agent-trace
 *
 * Fields mapped to null (documented as TODO:SPEC to match the Python side):
 *   vcs          — commit SHA not captured at insertion time
 *   tool.version — capturing tool version not available in extension
 *   content_hash — per-range code hash not computed
 *   url          — conversationId is not a URL
 */

import { CaptureRecord } from './store';

/** cursor/agent-trace spec version */
export const AGENT_TRACE_SPEC_VERSION = '0.1.0';
/** Internal LineageLens schema tag, stored in metadata for backward compat. */
export const LINEAGELENS_SCHEMA_VERSION = 'lineagelens-agent-trace/1';

/** Derive contributor.type from confidence score (spec values are all lowercase). */
function deriveContributorType(
  confidence: number | undefined,
  hasModel: boolean,
): 'human' | 'ai' | 'mixed' | 'unknown' {
  if (!hasModel) { return 'unknown'; }
  if (confidence === undefined) { return 'unknown'; }
  if (confidence >= 0.7) { return 'ai'; }
  if (confidence >= 0.3) { return 'mixed'; }
  return 'unknown';
}

/** Truncate inserted code to 120 chars and replace newlines (same as Python). */
function buildPreview(insertedCode: string): string | null {
  if (!insertedCode) { return null; }
  return insertedCode.slice(0, 120).replace(/\n/g, '↵');
}

/** Build one AgentTraceDocument JSON object from a CaptureRecord. */
export function captureToAgentTraceDoc(
  record: CaptureRecord,
  workspaceId: string,
  exportedAt: string,
): Record<string, unknown> {
  const contributorType = deriveContributorType(record.confidence, false);
  // Easy Mode has no tool/model data; contributor.type is based on source.
  // source === 'ai' means a known AI extension was active during insertion.
  const effectiveType: 'human' | 'ai' | 'mixed' | 'unknown' =
    record.source === 'ai'
      ? deriveContributorType(record.confidence, true)
      : deriveContributorType(record.confidence, false);

  // cursor_line is not available in Easy Mode; start_line defaults to 1.
  const startLine = 1;
  const endLine = Math.max(1, record.linesAdded);

  const contributor: Record<string, unknown> = { type: effectiveType };
  // model_id: Easy Mode cannot identify the model; omit rather than invent.
  // TODO:SPEC: populate model_id when the proxy or Copilot API exposes it.

  const metadata: Record<string, unknown> = {
    'lineagelens.schemaVersion': LINEAGELENS_SCHEMA_VERSION,
    'lineagelens.workspaceId': workspaceId,
    'lineagelens.exportedAt': exportedAt,
    'lineagelens.confidence': { score: record.confidence, level: null },
  };
  const preview = buildPreview(record.insertedCode);
  if (preview !== null) {
    metadata['lineagelens.insertedCodePreview'] = preview;
  }
  metadata['lineagelens.netAddedLines'] = record.linesAdded;

  const doc: Record<string, unknown> = {
    version: AGENT_TRACE_SPEC_VERSION,
    id: record.id,
    timestamp: record.timestamp,
    // vcs: TODO:SPEC — not captured at insertion time in Easy Mode
    files: [
      {
        path: record.filePath,
        conversations: [
          {
            // url: TODO:SPEC — no conversation URL available in Easy Mode
            contributor,
            ranges: [{ start_line: startLine, end_line: endLine }],
          },
        ],
      },
    ],
    metadata,
  };

  return doc;
}

/** Serialize an array of CaptureRecords to Agent Trace JSONL (one doc per line). */
export function captureStoreTojsonl(
  records: CaptureRecord[],
  workspaceId: string,
): string {
  const exportedAt = new Date().toISOString();
  const lines = records.map(r =>
    JSON.stringify(captureToAgentTraceDoc(r, workspaceId, exportedAt)),
  );
  return lines.join('\n') + (lines.length > 0 ? '\n' : '');
}
